include/boost/url/detail/impl/format_args.ipp

```diff
@@ -0,0 +1,642 @@
+            case '^':
+                lpad = pad / 2;
+                rpad = pad - lpad;
+                break;
+            }
+        }
+    }
+
+    // write
+    v = v0;
+    char* out = ctx.out();
+    if (!zeros)
+    {
+        for (std::size_t i = 0; i < lpad; ++i)
+            encode_one(out, fill, cs);
+    }
+    if (v < 0)
+    {
+        encode_one(out, '-', cs);
+        v *= -1;
+        --n;
+    }
+    else if (sign != '-')
+    {
+        encode_one(out, sign, cs);
+        --n;
+    }
+    if (zeros)
+    {
+        for (std::size_t i = 0; i < lpad; ++i)
+            encode_one(out, '0', cs);
+    }
+    while (n)
+    {
+        unsigned long long int d = v / p;
+        encode_one(out, '0' + static_cast<char>(d), cs);
+        --n;
+        v %= p;
+        p /= 10;
+    }
+    if (!zeros)
+    {
+        for (std::size_t i = 0; i < rpad; ++i)
+            encode_one(out, fill, cs);
+    }
+    return out;
+}
+
+char*
+integer_formatter_impl::
+format(
+unsigned long long int v,
+format_context& ctx,
+grammar::lut_chars const& cs) const
+{
+    // get n digits
+    unsigned long long int v0 = v;
+    unsigned long long int p = 1;
+    std::size_t n = 0;
+    if (sign != '-')
+    {
+        ++n;
+    }
+    while (v > 0)
+    {
+        if (v >= 10)
+            p *= 10;
+        v /= 10;
+        ++n;
+    }
+    static constexpr auto m =
+        std::numeric_limits<unsigned long long int>::digits10;
+    BOOST_ASSERT(n <= m + 1);
+    ignore_unused(m);
+
+    // get pad
+    std::size_t w = width;
+    if (width_idx != std::size_t(-1) ||
+        !width_name.empty())
+    {
+        get_width_from_args(
+            width_idx, width_name, ctx.args(), w);
+    }
+    std::size_t lpad = 0;
+    std::size_t rpad = 0;
+    if (w > n)
+    {
+        std::size_t pad = w - n;
+        if (zeros)
+        {
+            lpad = pad;
+        }
+        else
+        {
+            switch (align)
+            {
+            case '<':
+                rpad = pad;
+                break;
+            case '>':
+                lpad = pad;
+                break;
+            case '^':
+                lpad = pad / 2;
+                rpad = pad - lpad;
+                break;
+            }
+        }
+    }
+
+    // write
+    v = v0;
+    char* out = ctx.out();
+    if (!zeros)
+    {
+        for (std::size_t i = 0; i < lpad; ++i)
+            encode_one(out, fill, cs);
+    }
+    if (sign != '-')
+    {
+        encode_one(out, sign, cs);
+        --n;
+    }
+    if (zeros)
+    {
+        for (std::size_t i = 0; i < lpad; ++i)
+            encode_one(out, '0', cs);
+    }
+    while (n)
+    {
+        unsigned long long int d = v / p;
+        encode_one(out, '0' + static_cast<char>(d), cs);
+        --n;
+        v %= p;
+        p /= 10;
+    }
+    if (!zeros)
+    {
+        for (std::size_t i = 0; i < rpad; ++i)
+            encode_one(out, fill, cs);
+    }
+    return out;
+}
+
+} // detail
+} // urls
+} // boost
```

Comment:
This is a medium sized file, but at least its in the .ipp and not a visible header.